File: scala/difference-of-squares/src/main/scala/Squares.scala
object Squares {
  val One = BigInt(1)

  def squareOfSums =
    (One to (_:BigInt)) andThen
      (_.sum) andThen
      (_.pow(2))

  def sumOfSquares =
    (One to (_:BigInt)) andThen
      (_.map(_.pow(2))) andThen
      (_.sum)

  def difference(i: BigInt) =
    squareOfSums(i) - sumOfSquares(i)

  def apply() = Squares
}

File: haskell/difference-of-squares/src/Squares.hs
module Squares (difference, squareOfSum, sumOfSquares) where

difference :: Integral a => a -> a
difference n =  (squareOfSum n) - (sumOfSquares n)

squareOfSum :: Integral a => a -> a
squareOfSum n = square . sum $ [1..n]

sumOfSquares :: Integral a => a -> a
sumOfSquares n =  sum $ map square [1..n]

square :: Num a => a -> a
square = (^ (2 :: Int))
